internal/controller/user_controller.go

```diff
@@ -0,0 +1,368 @@
+	newPassword, err := GenerateRandomPassword(16)
+	if err != nil {
+		return fmt.Errorf("failed to generate new password for BMC account %s: %w", user.Name, err)
+	}
+	newSecret, err := r.createSecret(ctx, log, user, string(newPassword))
+	if err != nil {
+		return fmt.Errorf("failed to create BMCSecret: %w", err)
+	}
+	log.Info("Creating BMC account with new password", "Account", user.Name)
+	if err = r.updatePassword(ctx, log, user, string(newPassword), newSecret, bmcClient); err != nil {
+		return fmt.Errorf("failed to update BMC account password: %w", err)
+	}
+	log.Info("BMC account created with new password", "Account", user.Name)
+
+	return nil
+}
+
+func (r *UserReconciler) handleMissingEffectiveBMCSecretRef(ctx context.Context, log logr.Logger, bmcClient bmc.BMC, bmcSecret *metalv1alpha1.BMCSecret, user *metalv1alpha1.User) (ctrl.Result, error) {
+	log.Info("No effective BMCSecret reference set for User", "User", user.Name)
+	_, password, err := bmcutils.GetBMCCredentialsFromSecret(bmcSecret)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get credentials from BMCSecret: %w", err)
+	}
+	if err = r.updatePassword(ctx, log, user, password, bmcSecret, bmcClient); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to update BMC account password: %w", err)
+	}
+	log.Info("Created effective BMCSecret for User", "User", user.Name)
+	if user.Spec.RotationPolicy == nil {
+		// If no rotation period is set, we don't need to requeue
+		return ctrl.Result{}, nil
+	}
+	log.Info("Requeuing for password rotation", "User", user.Name, "RotationPolicy", user.Spec.RotationPolicy.Duration)
+	// If a rotation period is set, we requeue for the next rotation
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: user.Spec.RotationPolicy.Duration,
+	}, nil
+}
+
+func (r *UserReconciler) handleUpdatedSecretRef(ctx context.Context, log logr.Logger, user *metalv1alpha1.User, bmcSecret *metalv1alpha1.BMCSecret, bmcClient bmc.BMC) error {
+	log.Info("BMCSecret credentials have changed, updating BMC user", "User", user.Name)
+	_, password, err := bmcutils.GetBMCCredentialsFromSecret(bmcSecret)
+	if err != nil {
+		return fmt.Errorf("failed to get credentials from BMCSecret: %w", err)
+	}
+	if err = r.updatePassword(ctx, log, user, password, bmcSecret, bmcClient); err != nil {
+		return fmt.Errorf("failed to update BMC account password: %w", err)
+	}
+	return nil
+}
+
+func (r *UserReconciler) updatePassword(ctx context.Context, log logr.Logger, user *metalv1alpha1.User, password string, secret *metalv1alpha1.BMCSecret, bmcClient bmc.BMC) error {
+	if err := r.removeEffectiveSecret(ctx, log, user); err != nil {
+		return fmt.Errorf("failed to remove effective BMCSecret: %w", err)
+	}
+	// Update the BMC account with the new password
+	if err := bmcClient.CreateOrUpdateAccount(ctx, user.Spec.UserName, user.Spec.RoleID, password, r.Insecure); err != nil {
```

Comment:
Don't we loose the password, if this call fails?

Reply:
done